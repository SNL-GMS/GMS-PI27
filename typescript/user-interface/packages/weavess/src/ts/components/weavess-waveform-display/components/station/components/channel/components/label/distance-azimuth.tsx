/* eslint-disable react/jsx-no-useless-fragment */
import { Tooltip } from '@blueprintjs/core';
import { WeavessTypes } from '@gms/weavess-core';
import React from 'react';

/**
 * The type of the props for the {@link DistanceAzimuth} component
 */
export interface DistanceAzimuthProps {
  /** Distance */
  distance: number;

  /** Distance units */
  distanceUnits: WeavessTypes.DistanceUnits;

  /** Azimuth */
  azimuth: number;
}

/**
 * Creates a distance/azimuth value for the label with wrapped tooltips and classes for
 * styling. Use km vs degree ('\u00B0') symbol depending on distanceUnits enum from props
 *
 */
export function InternalDistanceAzimuth(props: DistanceAzimuthProps) {
  const { distanceUnits, distance, azimuth } = props;
  const disFixBy = distanceUnits === WeavessTypes.DistanceUnits.degrees ? 1 : 2;

  return (
    <>
      {distance !== 0 && (
        <Tooltip className="label-tooltip-wrapper__value" content={`Distance (${distanceUnits})`}>
          <>
            {distance.toFixed(disFixBy)}
            {distanceUnits === WeavessTypes.DistanceUnits.km ? ' km' : '\u00B0'}
          </>
        </Tooltip>
      )}
      {azimuth !== 0 && (
        <>
          /
          <Tooltip className="label-tooltip-wrapper__value" content="Azimuth (degrees)">
            <>{`${azimuth.toFixed(1)}\u00B0`}</>
          </Tooltip>
        </>
      )}
    </>
  );
}

export const DistanceAzimuth = React.memo(InternalDistanceAzimuth);
