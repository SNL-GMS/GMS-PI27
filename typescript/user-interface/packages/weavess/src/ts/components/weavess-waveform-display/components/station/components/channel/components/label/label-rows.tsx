import { Tooltip } from '@blueprintjs/core';
import type { DistanceUnits } from '@gms/weavess-core/lib/types';
import classNames from 'classnames';
import React from 'react';

import { DistanceAzimuth } from './distance-azimuth';

export interface ChooseWaveformRowProps {
  isDefaultChannel: boolean;
  isSplitChannel: boolean;
}

export interface ChannelNameRowProps {
  isSplitChannel: boolean;
  channelLabel?: string;
  channelLabelTooltip?: string;
  channelName: string | JSX.Element;
  phaseLabel: string | undefined;
  phaseColor: string | undefined;
  isDefaultChannel: boolean;
  tooltipText?: string;
}
export interface PhaseLabelProps {
  phaseLabel: string;
  phaseColor: string | undefined;
  tooltipText?: string;
}

export interface AzimuthMaskRowProps {
  maskIndicator: string | null;
  azimuth: number;
  distance: number;
  distanceUnits: DistanceUnits;
}

export interface LabelHeaderRowProps {
  labelHeader: string | JSX.Element;
}

export interface FKRowProps {
  id: string;
}

/**
 * Special formatting for FK channel label rows
 *
 * @param props uses ID to determine FK type
 * @returns name row for FK plots
 */
export const FKNameRow = React.memo(function FKNameRow(props: FKRowProps) {
  const { id } = props;
  if (id === 'Slowness') {
    return (
      <div style={{ whiteSpace: 'nowrap' }}>
        Slowness (<sup>s</sup>
        &#8725;
        <sub>&deg;</sub>)
      </div>
    );
  }
  if (id === 'Azimuth') {
    return (
      <div style={{ whiteSpace: 'nowrap' }}>
        Azimuth <sup>(&deg;)</sup>
      </div>
    );
  }
  return null;
});

const PhaseLabel = React.memo(function PhaseLabel({
  tooltipText,
  phaseLabel,
  phaseColor
}: PhaseLabelProps) {
  const phaseColorStyle = {
    color: phaseColor
  };
  return tooltipText ? (
    <Tooltip content={tooltipText}>
      <span className="station-name__phase-name" style={phaseColorStyle}>
        {`/${phaseLabel}`}
      </span>
    </Tooltip>
  ) : (
    <span className="station-name__phase-name" style={phaseColorStyle}>
      {`/${phaseLabel}`}
    </span>
  );
});

/**
 * Exists only when split channel mode expansion is active, and only for Default Channels
 *
 * @param isDefaultChannel
 * @param isSplitChannel
 * @returns null when not a default channel in split channel expanded mode
 */
export const ChooseWaveformRow = React.memo(function ChooseWaveformRow(
  props: ChooseWaveformRowProps
) {
  const { isDefaultChannel, isSplitChannel } = props;
  return isSplitChannel && isDefaultChannel ? <span>CHOOSE WAVEFORM</span> : null;
});

/**
 * Includes the channel name.
 * For split non-default channels, also contains Signal Detection phase colored to match event association
 *
 * @param props
 * @returns a row for weavess channel labels
 */
export const ChannelNameRow = React.memo(function ChannelNameRow(props: ChannelNameRowProps) {
  const {
    isSplitChannel,
    channelName,
    channelLabel,
    phaseColor,
    phaseLabel,
    isDefaultChannel,
    tooltipText
  } = props;
  // For displaying labels in FK plots
  if (channelName === 'Slowness' || channelName === 'Azimuth') {
    return <FKNameRow id={channelName} />;
  }

  return (
    <div className="label-channel-name-row">
      <span
        className={classNames(
          { 'station-name__channel-name': !isDefaultChannel },
          { 'station-name': isDefaultChannel }
        )}
      >
        {channelName}
      </span>
      <span className="station-name__channel-name">&nbsp;{channelLabel}</span>
      {isSplitChannel && !isDefaultChannel && phaseLabel ? (
        <PhaseLabel tooltipText={tooltipText} phaseLabel={phaseLabel} phaseColor={phaseColor} />
      ) : null}
    </div>
  );
});

/**
 *
 * @param props Includes the Azimuth and Distance when an event is open, as well as an indicator if a mask is applied
 * @returns a row for weavess channel labels
 */
export const AzimuthMaskRow = React.memo(function AzimuthMaskRow(props: AzimuthMaskRowProps) {
  const { maskIndicator, azimuth, distance, distanceUnits } = props;
  return (
    <div className="distance-azimuth">
      <p>
        <DistanceAzimuth azimuth={azimuth} distance={distance} distanceUnits={distanceUnits} />
        <span className="label-container-content-mask-indicator">{maskIndicator}</span>
      </p>
    </div>
  );
});

/**
 * @returns simple row to display a label header
 */
export const LabelHeaderRow = React.memo(function LabelHeaderRow(props: LabelHeaderRowProps) {
  const { labelHeader } = props;

  return <div className="label-header">{labelHeader}</div>;
});
