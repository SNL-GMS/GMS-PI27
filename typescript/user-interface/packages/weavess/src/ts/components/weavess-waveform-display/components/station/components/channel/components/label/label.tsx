/* eslint-disable react/destructuring-assignment */
import { Tooltip } from '@blueprintjs/core';
import { classList, UILogger } from '@gms/ui-util';
import type { WeavessTypes } from '@gms/weavess-core';
import React from 'react';

import { YAxis } from '../../../../../axes';
import { LabelLeftElement } from './label-elements';
import { AzimuthMaskRow, ChannelNameRow, ChooseWaveformRow, LabelHeaderRow } from './label-rows';
import type { LabelState } from './types';

const logger = UILogger.create('GMS_LOG_WEAVESS', process.env.GMS_LOG_WEAVESS);

/**
 * Label component. Describes a waveform (or other graphic component) and has optional events
 */
export class Label extends React.PureComponent<WeavessTypes.LabelProps, LabelState> {
  /** The y-axis references. */
  public yAxisRefs: { [id: string]: YAxis | null } = {};

  /**
   * Constructor
   *
   * @param props Label props as LabelProps
   */
  public constructor(props: WeavessTypes.LabelProps) {
    super(props);
    this.state = {};
  }

  // ******************************************
  // BEGIN REACT COMPONENT LIFECYCLE METHODS
  // ******************************************

  /**
   * Called immediately after updating occurs. Not called for the initial render.
   *
   * @param prevProps the previous props
   * @param prevState the previous state
   */
  public componentDidUpdate(prevProps: WeavessTypes.LabelProps): void {
    // should we force re-draw the yAxis?
    // Note that the y-axis will redraw based on its changed props,
    // so we don't need to check for changes to the bounds
    if (
      this.props.channel.id !== prevProps.channel.id ||
      this.props.channel?.height !== prevProps.channel?.height ||
      this.props.channel.yAxisTicks !== prevProps.channel.yAxisTicks
    ) {
      this.refreshYAxis();
    }
  }

  /**
   * Catches exceptions generated in descendant components.
   * Unhandled exceptions will cause the entire component tree to unmount.
   *
   * @param error the error that was caught
   * @param info the information about the error
   */
  public componentDidCatch(error, info): void {
    logger.error(`Weavess Label Error: ${error} : ${info}`);
  }

  // ******************************************
  // END REACT COMPONENT LIFECYCLE METHODS
  // ******************************************

  /**
   * Creates the phase label and color string based on props.
   * If all are the same, returns that value. Otherwise, returns `*` for the label and
   * undefined for the color.
   *
   * @returns a phase label string and color string
   */
  private readonly getPhaseLabelAndColor = (): {
    phaseLabel: string | undefined;
    phaseColor: string | undefined;
  } => {
    if (this.props.channel?.waveform?.signalDetections) {
      const phaseLabel = this.props.channel.waveform.signalDetections.reduce<string | undefined>(
        (label, phase) => {
          return label === phase.label ? label : '*';
        },
        this.props.channel.waveform.signalDetections[0]?.label
      );
      const phaseColor = this.props.channel.waveform.signalDetections.reduce<string | undefined>(
        (color, phase) => {
          return color === phase.color ? color : undefined;
        },
        this.props.channel.waveform.signalDetections[0]?.label
      );
      return { phaseLabel, phaseColor };
    }
    return { phaseColor: undefined, phaseLabel: undefined };
  };

  // eslint-disable-next-line react/sort-comp
  public render(): JSX.Element {
    const { phaseLabel, phaseColor } = this.getPhaseLabelAndColor();
    const maskIndicator = this.props.showMaskIndicator ? 'M' : null;
    return (
      <div
        onContextMenu={this.onContextMenu}
        // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
        tabIndex={0}
        className={`label${this.props?.events?.onChannelLabelClick ? ' label--actionable' : ''}`}
      >
        <div className={`label-container${this.props.channel.isSelected ? ' is-selected' : ''}`}>
          <LabelLeftElement
            isDefaultChannel={this.props.isDefaultChannel}
            isExpandable={this.props.isExpandable}
            isSplitChannel={!!this.props.channel.splitChannelTime}
            isExpanded={this.props.expanded}
            labelContainerOnClick={this.labelContainerOnClick}
            closeSplitChannelOverlayCallback={this.props.closeSplitChannelOverlayCallback}
            channelId={this.props.channel.id}
          />
          <div className="label-container-content">
            {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */}
            <div
              className={`label-container-content-label${
                this.props.channel.isSelected ? ' is-selected' : ''
              }`}
              onClick={this.labelContainerContentOnClick}
              data-cy={`channel-label-${this.props.channel.id}`}
            >
              {this.props.customLabel ? (
                // eslint-disable-next-line react/jsx-props-no-spreading
                <this.props.customLabel {...this.props} />
              ) : (
                <div>
                  <span className="weavess-tooltip__target label-container-content">
                    {!this.props.isDefaultChannel && this.props.labelHeader && (
                      <LabelHeaderRow labelHeader={this.props.labelHeader} />
                    )}
                    <Tooltip
                      disabled={!this.props.channelLabelTooltip}
                      className={classList({
                        'weavess-tooltip': true,
                        'weavess-tooltip--help': !!this.props.channelLabelTooltip
                      })}
                      content={this.props.channelLabelTooltip}
                      placement="right"
                      hoverOpenDelay={250} // ms
                    >
                      <span className="station-label-container">
                        <ChooseWaveformRow
                          isDefaultChannel={this.props.isDefaultChannel}
                          isSplitChannel={!!this.props.channel.splitChannelTime}
                        />
                        <ChannelNameRow
                          isSplitChannel={!!this.props.channel.splitChannelTime}
                          isDefaultChannel={this.props.isDefaultChannel}
                          channelName={this.props.channel.name}
                          channelLabel={this.props.channel.channelLabel}
                          channelLabelTooltip={this.props.channelLabelTooltip}
                          phaseColor={phaseColor}
                          phaseLabel={phaseLabel}
                          tooltipText={phaseLabel === '*' ? 'Multiple phases' : undefined}
                        />
                      </span>
                    </Tooltip>
                    <AzimuthMaskRow
                      maskIndicator={maskIndicator}
                      azimuth={this.props.azimuth}
                      distance={this.props.distance}
                      distanceUnits={this.props.distanceUnits}
                    />
                  </span>
                </div>
              )}
            </div>
            <div
              style={{
                height: '100%'
              }}
              data-cy-contains-amplitude-markers={this.hasYAxes()}
            >
              {this.props.suppressLabelYAxis ||
              !this.props.yAxisBounds?.waveformYAxisBounds ? undefined : (
                <YAxis
                  key={`${this.props.channel.id}_yaxis_waveform`}
                  ref={ref => {
                    this.yAxisRefs.waveformYAxisBounds = ref;
                  }}
                  maxAmplitude={this.props.yAxisBounds.waveformYAxisBounds?.maxAmplitude}
                  minAmplitude={this.props.yAxisBounds.waveformYAxisBounds?.minAmplitude}
                  heightInPercentage={
                    this.props.yAxisBounds.waveformYAxisBounds?.heightInPercentage
                  }
                  yAxisTicks={this.props.channel.yAxisTicks}
                />
              )}
              {this.props.suppressLabelYAxis ||
              !this.yAxisRefs?.spectrogramYAxisBounds ? undefined : (
                <YAxis
                  key={`${this.props.channel.id}_yaxis_spectrogram`}
                  ref={ref => {
                    this.yAxisRefs.spectrogramYAxisBounds = ref;
                  }}
                  maxAmplitude={this.props.yAxisBounds.spectrogramYAxisBounds?.maxAmplitude}
                  minAmplitude={this.props.yAxisBounds.spectrogramYAxisBounds?.minAmplitude}
                  heightInPercentage={
                    this.props.yAxisBounds.spectrogramYAxisBounds?.heightInPercentage
                  }
                  yAxisTicks={this.props.channel.yAxisTicks}
                />
              )}
            </div>
          </div>
        </div>
      </div>
    );
  }

  /**
   * Refreshes the y-axis for the label
   */
  public readonly refreshYAxis = (): void => {
    if (this.yAxisRefs) {
      Object.keys(this.yAxisRefs).forEach(key => {
        const yAxis = this.yAxisRefs[key];
        if (yAxis) {
          yAxis.display();
        }
      });
    }
  };

  private readonly hasYAxes = (): boolean => {
    return (
      !!this.props.yAxisBounds?.waveformYAxisBounds ||
      !!this.props.yAxisBounds?.spectrogramYAxisBounds
    );
  };

  /**
   * Helper method to reduce code complexity
   *
   * @param amplitude
   * @returns
   */
  // eslint-disable-next-line class-methods-use-this
  private readonly amplitudeOrFallback = (amplitude: number) => {
    return amplitude ?? -1;
  };

  /**
   * The on context menu event handler
   * Does not fire if channel is in split-expansion mode
   *
   * @param e the mouse event
   */
  private readonly onContextMenu = (e: React.MouseEvent<HTMLDivElement, MouseEvent>): void => {
    if (this.props.events?.onContextMenu && (e.button === 2 || (e.button === 0 && e.ctrlKey))) {
      e.preventDefault();
      if (this.props.channel.splitChannelTime) return;
      this.props.events.onContextMenu(
        e,
        this.props.channelName,
        this.hasYAxes() && this.props.yAxisBounds.waveformYAxisBounds?.minAmplitude != null
          ? this.amplitudeOrFallback(this.props.yAxisBounds.waveformYAxisBounds?.minAmplitude)
          : -1,
        this.hasYAxes() && this.props.yAxisBounds.waveformYAxisBounds?.maxAmplitude != null
          ? this.amplitudeOrFallback(this.props.yAxisBounds.waveformYAxisBounds?.maxAmplitude)
          : -1,
        this.props.isDefaultChannel,
        this.props.isMeasureWindow
      );
    }

    // Prevents chrome context menu from appearing if
    // Left mouse + control is used to summon context menu
    if (this.props.events?.onChannelLabelClick && e.button === 0 && e.ctrlKey) {
      e.preventDefault();
      e.stopPropagation();
      if (this.props.channel.splitChannelTime) return;
      this.props.events.onChannelLabelClick(e, this.props.channel.id);
    }
  };

  /**
   * The label container on click event handler
   *
   * @param e the mouse event
   */
  private readonly labelContainerOnClick = () => {
    const isExpanded = !this.props.expanded;
    if (isExpanded) {
      if (this.props.events && this.props.events.onChannelExpanded) {
        this.props.events.onChannelExpanded(this.props.channel.id);
      }
    } else if (this.props.events && this.props.events.onChannelCollapsed) {
      this.props.events.onChannelCollapsed(this.props.channel.id);
    }
  };

  /**
   * The label container content on click event handler
   *
   * @param e the mouse event
   */
  private readonly labelContainerContentOnClick = (
    e: React.MouseEvent<HTMLDivElement, MouseEvent>
  ) => {
    if (this.props.events && this.props.events.onChannelLabelClick) {
      this.props.events.onChannelLabelClick(e, this.props.channel.id);
    }
  };
}
