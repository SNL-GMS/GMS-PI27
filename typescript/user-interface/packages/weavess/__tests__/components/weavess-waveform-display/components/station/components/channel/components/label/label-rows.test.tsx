import { queryByText, render } from '@testing-library/react';
import * as React from 'react';

import { ChannelNameRow } from '../../../../../../../../../src/ts/components/weavess-waveform-display/components/station/components/channel/components/label/label-rows';

describe('RowLabel', () => {
  it('renders a label for default channel given a channelLabel', () => {
    const { container } = render(
      <ChannelNameRow
        channelName="ASAR"
        channelLabel="beam.SHZ"
        isDefaultChannel
        isSplitChannel={false}
        phaseColor="FFF"
        phaseLabel="PP"
      />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText(container, 'ASAR')).not.toBeNull();
    expect(queryByText(container, 'beam.SHZ')).not.toBeNull();
    // Not in split mode, and is default channel
    expect(queryByText(container, 'PP')).toBeNull();
  });

  it('renders a label for non-default channel in split mode with phase label', () => {
    const { container } = render(
      <ChannelNameRow
        channelName="ASAR"
        channelLabel="beam.SHZ"
        isDefaultChannel={false}
        isSplitChannel
        phaseColor="FFF"
        phaseLabel="PP"
      />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText(container, 'ASAR')).not.toBeNull();
    expect(queryByText(container, 'beam.SHZ')).not.toBeNull();
    // In split mode, non-default channel
    expect(queryByText(container, '/PP')).not.toBeNull();
  });

  it('renders a label for default channel in split mode without phase label', () => {
    const { container } = render(
      <ChannelNameRow
        channelName="ASAR"
        channelLabel="beam.SHZ"
        isDefaultChannel
        isSplitChannel
        phaseColor="FFF"
        phaseLabel="PP"
      />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText(container, 'ASAR')).not.toBeNull();
    expect(queryByText(container, 'beam.SHZ')).not.toBeNull();
    // In split mode, but is a default channel
    expect(queryByText(container, '/PP')).toBeNull();
  });
  it('renders a label for default channel without a channelLabel (normally for station with no data)', () => {
    const { container } = render(
      <ChannelNameRow
        channelName="ASAR"
        channelLabel=""
        isDefaultChannel
        isSplitChannel={false}
        phaseColor="FFF"
        phaseLabel="PP"
      />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText(container, 'ASAR')).not.toBeNull();
  });

  it('creates labels for signal detections on a single raw channel', () => {
    const { container } = render(
      <ChannelNameRow
        channelName="ASAR.AS01.SHZ"
        channelLabel=""
        isDefaultChannel={false}
        isSplitChannel={false}
        phaseColor="FFF"
        phaseLabel="PP"
      />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText(container, 'ASAR.AS01.SHZ')).toBeDefined();
  });

  it('correctly generates Slowness label for FK display', () => {
    const { container } = render(
      <ChannelNameRow
        channelName="Slowness"
        channelLabel="oaeuaoeu"
        isDefaultChannel={false}
        isSplitChannel={false}
        phaseColor="FFF"
        phaseLabel="PP"
      />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText(container, '°')).not.toBeNull();
  });

  it('correctly generates Azimuth label for FK display', () => {
    const { container } = render(
      <ChannelNameRow
        channelName="Azimuth"
        channelLabel="oaeuaoeu"
        isDefaultChannel={false}
        isSplitChannel={false}
        phaseColor="FFF"
        phaseLabel="PP"
      />
    );
    expect(container).toMatchSnapshot();
    expect(queryByText(container, '(°)')).not.toBeNull();
  });
});
