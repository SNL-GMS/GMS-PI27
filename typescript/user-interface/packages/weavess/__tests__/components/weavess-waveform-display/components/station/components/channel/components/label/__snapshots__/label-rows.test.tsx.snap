// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RowLabel correctly generates Azimuth label for FK display 1`] = `
<div>
  <div
    style="white-space: nowrap;"
  >
    Azimuth 
    <sup>
      (°)
    </sup>
  </div>
</div>
`;

exports[`RowLabel correctly generates Slowness label for FK display 1`] = `
<div>
  <div
    style="white-space: nowrap;"
  >
    Slowness (
    <sup>
      s
    </sup>
    ∕
    <sub>
      °
    </sub>
    )
  </div>
</div>
`;

exports[`RowLabel creates labels for signal detections on a single raw channel 1`] = `
<div>
  <div
    class="label-channel-name-row"
  >
    <span
      class="station-name__channel-name"
    >
      ASAR.AS01.SHZ
    </span>
    <span
      class="station-name__channel-name"
    >
       
    </span>
  </div>
</div>
`;

exports[`RowLabel renders a label for default channel given a channelLabel 1`] = `
<div>
  <div
    class="label-channel-name-row"
  >
    <span
      class="station-name"
    >
      ASAR
    </span>
    <span
      class="station-name__channel-name"
    >
       
      beam.SHZ
    </span>
  </div>
</div>
`;

exports[`RowLabel renders a label for default channel in split mode without phase label 1`] = `
<div>
  <div
    class="label-channel-name-row"
  >
    <span
      class="station-name"
    >
      ASAR
    </span>
    <span
      class="station-name__channel-name"
    >
       
      beam.SHZ
    </span>
  </div>
</div>
`;

exports[`RowLabel renders a label for default channel without a channelLabel (normally for station with no data) 1`] = `
<div>
  <div
    class="label-channel-name-row"
  >
    <span
      class="station-name"
    >
      ASAR
    </span>
    <span
      class="station-name__channel-name"
    >
       
    </span>
  </div>
</div>
`;

exports[`RowLabel renders a label for non-default channel in split mode with phase label 1`] = `
<div>
  <div
    class="label-channel-name-row"
  >
    <span
      class="station-name__channel-name"
    >
      ASAR
    </span>
    <span
      class="station-name__channel-name"
    >
       
      beam.SHZ
    </span>
    <span
      class="station-name__phase-name"
    >
      /PP
    </span>
  </div>
</div>
`;
